File: class/1264/A-Z/Y.c
/*
User: shenyouran
Problem: 1336
Algorithm: DFS
Language: C/C++
Record ID: 74614
Submit Time: 2020-10-18 10:10:35
*/
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<stdbool.h>
int v,g,a[26],b[16][26],id[16];
bool f[16];
void dfs(int k,int x)
{
	if(k>x)
	{
		int c[26]={};
		for(int i=1;i<=x;i++)for(int j=1;j<=v;j++)c[j]+=b[id[i]][j];
		for(int i=1;i<=v;i++)if(a[i]>c[i])return;
		printf("%d ",x);
		for(int i=1;i<=x;i++)printf("%d ",id[i]);
		exit(0);
	}
	for(int i=id[k-1]+1;i<=g;i++)
	{
		if(f[i])continue;
		id[k]=i;
		f[i]=true;
		dfs(k+1,x);
		f[i]=false;
	}
}
int main()
{
	scanf("%d",&v);
	for(int i=1;i<=v;i++)scanf("%d",&a[i]);
	scanf("%d",&g);
	for(int i=1;i<=g;i++)for(int j=1;j<=v;j++)scanf("%d",&b[i][j]);
	for(int i=1;i<=g;i++,memset(id,0,sizeof(id)),memset(f,false,sizeof(f)))dfs(1,i);
	return 0;
}

File: class/1264/uppercase/Y.pas
(*
User: shenyouran
Problem: 1336
Algorithm: DFS
Language: Pascal
Record ID: 74615
Submit Time: 2020-10-18 10:15:07
*)
var
    v,g,i,j:integer;
    a:array[1..25] of integer;
    b:array[1..15,1..25] of integer;
    id:array[1..15] of integer;
    f:array[1..15] of boolean;
procedure dfs(k,x:integer);
    var
        i:integer;
        c:array[1..25] of longint;
    begin
        if k>x then
            begin
                fillchar(c,sizeof(c),0);
                for i:=1 to x do for j:=1 to v do inc(c[j],b[id[i]][j]);
                for i:=1 to v do if a[i]>c[i] then exit;
                write(x,' ');
                for i:=1 to x do write(id[i],' ');
                halt;
            end;
        for i:=id[k-1]+1 to g do
            begin
                if f[i] then continue;
                id[k]:=i;
                f[i]:=true;
                dfs(k+1,x);
                f[i]:=false;
            end;
    end;
begin
    read(v);
    for i:=1 to v do read(a[i]);
    read(g);
    for i:=1 to g do for j:=1 to v do read(b[i][j]);
    for i:=1 to g do
        begin
            for j:=1 to v do
                begin
                    fillchar(id,sizeof(id),0);
                    fillchar(f,sizeof(f),false);
                    dfs(1,i);
                end;
        end;
end.